### ISSUE TITLE
Discussion: a validity mask for camera projections and depth warps (z ≤ eps, missing depth, points behind the camera)

### ISSUE THREAD

**Issue by ducha-aiki** [kornia/kornia#4555]
**This is a discussion issue.** It asks for a decision; there is no fix to review yet.

#4267 is being settled for the projected *coordinates*: below `abs(z) <= eps` the divide is skipped, and the result stays finite, gradient-safe and compile-safe (#4475 implements this). That rule gives every point a pixel, including points that have no meaningful one. What is still missing is a way to tell a caller *which* projections are valid. Kornia has no shared answer today, and the depth warps show the cost.

## What happens without a validity signal

Measured on `main` @ `1590df48`, torch 2.14.0, CPU:

1. **Missing depth (encoded as `0` by most depth sensors) samples a real pixel, and the two warps pick different ones.** Setup: a 4x5 `arange` image, `K = [[2, 0, 2], [0, 2, 1], [0, 0, 1]]`, identity pose, depth `1` except `0` at `(0,0)`, `(2,1)` and `(3,4)`.
   - `warp_frame_depth` fills all three holes with `8.0`, the pixel at the principal point.
   - `DepthWarper` fills the same holes with `1.0`, pixel `(0, 0)`.
2. **Points behind the camera are sampled as if they were in front.** `warp_frame_depth` with unit depth and a `z = -2` translation, so every transformed point has `z = -1`, returns the image mirrored about the principal point. 15 of 20 output pixels are sampled image values, and the other 5 fall outside the image.
3. **`inf` cannot be the validity signal on our torch floor.** On torch 2.5.1 (arm64 CPU), `F.grid_sample` segfaults (exit 139) on a grid holding `inf`, as it already does for `1e12` (pytorch/pytorch#24823; this is why the #4267 pin in `tests/geometry/test_depth.py` no longer samples through `DepthWarper`'s grid). `nan` coordinates return `nan` on both 2.5.1 and 2.14. So "return a non-finite pixel and let the caller check `isfinite`" either crashes or poisons the output.
4. **Callers already build their own masks, with their own rules.** `kornia.feature.loftr.utils.geometry.warp_kpts` combines `depth != 0`, in-image covisibility and a 0.2 depth-consistency test, and it projects with `z + 1e-4`, a sixth `z` policy next to the five in #4267. `kornia.geometry.pointcloud` filters with `isfinite`.

<details><summary>Repro</summary>

```python
import torch
from kornia.geometry.camera import PinholeCamera
from kornia.geometry.depth import DepthWarper, warp_frame_depth

img = torch.arange(1, 21, dtype=torch.float32).view(1, 1, 4, 5)
K = torch.tensor([[[2.0, 0.0, 2.0], [0.0, 2.0, 1.0], [0.0, 0.0, 1.0]]])
depth = torch.ones(1, 1, 4, 5)
depth[0, 0, 0, 0] = depth[0, 0, 2, 1] = depth[0, 0, 3, 4] = 0  # missing depth

print(warp_frame_depth(img, depth, torch.eye(4)[None], K)[0, 0])
# holes -> 8.0 (the principal-point pixel)

K4 = torch.eye(4)[None]; K4[0, :3, :3] = K[0]
h, w = torch.tensor([4]), torch.tensor([5])
warper = DepthWarper(PinholeCamera(K4.clone(), torch.eye(4)[None], h, w), 4, 5)
warper.compute_projection_matrix(PinholeCamera(K4.clone(), torch.eye(4)[None], h, w))
print(warper(depth, img)[0, 0])
# holes -> 1.0 (pixel (0, 0))

T = torch.eye(4)[None]; T[0, 2, 3] = -2.0  # every point lands at z = -1
print(warp_frame_depth(img, torch.ones(1, 1, 4, 5), T, K)[0, 0])
# mirrored image, 15/20 pixels sampled
```

</details>

## Questions to decide

1. **What "valid" means.**
   - Candidate rule: `z > eps`, with `nan` invalid.
   - Is `eps` absolute (the `1e-8` of the coordinate rule, whose meaning depends on scene units) or dtype-aware?
   - Angle-based models such as `KannalaBrandtK3` legitimately project rays past 90°, so is validity a per-camera-model rule rather than a sign test?
   - Is in-image bounds (covisibility) part of validity, or a separate mask?
2. **API shape.** Options:
   - **(a)** a separate function or method returning a `bool` mask (for example `projection_valid_mask(points_cam, eps)`, and `CameraModelBase.valid(points)` for `kornia.sensors`). No signature change, scriptable, composable.
   - **(b)** a `return_valid: bool = False` flag on the projection functions. Convenient, but a return type that depends on a flag is awkward for TorchScript and ONNX export.
   - **(c)** masks only at the consumer level: `warp_frame_depth(..., return_mask=True)`, `DepthWarper` and `warp_kpts`.
   - **(d)** a sentinel coordinate (`nan`) instead of a mask. Ruled out by point 3 unless every sampler sanitizes first.
3. **Warp defaults.** Should `warp_frame_depth` and `DepthWarper` fill invalid pixels with padding (zeros) by default? That changes today's outputs in both cases above, and would be a breaking change for the coordinated window. The alternative is to fill only when asked, for example `return_mask=True` plus a fill value.
4. **Scope.** Which call sites adopt it:
   - `project_points`, `PinholeCamera.project`, `project_points_z1`, `cam2pixel`;
   - the `kornia.sensors` camera models;
   - the depth warps;
   - `warp_kpts` (which would also retire its `+ 1e-4`).
5. **Gradients.** The mask is non-differentiable. Should invalid points get exactly zero gradient through the coordinates? That needs a safe-argument `torch.where`, per `AGENTS.md`, so it holds on torch 2.5.1 as well as 2.14.

## A starting point, for discussion

- **Masks:** (a) for the geometric primitives, and (c) for the warps, as a `return_mask` flag that defaults to off, so nothing changes until the repair window.
- **Rule:** `z > eps` with `nan` invalid, overridden per camera model where the domain is angular.
- **Warps:** fill invalid pixels with padding, not the pass-through coordinate. This also removes the grid that crashes torch 2.5.1.
- **Later:** once this is agreed, flip the warp default in the coordinated window.

Related: #4267 (coordinate policy), #4475 (implements it), #4274 (`kornia.sensors` kept separate).

Posted on behalf of @ducha-aiki by Claude (Opus 5).

**Comment by NaCr05** [comment]
Thanks for the detailed review. I have converted #4701 to a draft. I agree that the PR selected several public policies before this discussion had settled them. I propose separating the pinhole predicate from the warp integration and agreeing on the following contract first.

**1. Pinhole eligibility and the coordinate policy.** I would keep the independent predicate, explicitly scoped to finite camera-frame points with `z > eps`. I found the later [#4267 decision](https://github.com/kornia/kornia/issues/4267#issuecomment-5678077316) specifying an absolute `1e-8` coordinate cutoff. I propose aligning the predicate with the cutoff of the projection it accompanies, rather than choosing a new dtype-dependent length threshold. An application's physical near plane can be an additional filter; machine epsilon alone does not express that distance. Half-precision comparison/working precision is a separate numerical decision.

This predicate would not certify projected-coordinate representability, image bounds or visibility. Angular and orthographic models need their own domain rules. I would leave those model-specific changes outside this first implementation. Also, #4475 is currently closed and unmerged, so the coordinate-policy decision and its implementation need to be tracked separately; this PR should not imply that all projection paths already agree.

**2. Bounds and filling.** I propose keeping geometric eligibility and image support separate internally, and excluding occlusion from this API: neither warp receives the additional depth needed to establish it. The current PR's centre-bounds policy is a real behavior change. In a local all-ones image example, sampling at `x=-0.25` yields `0.75` through the existing zero-padding interpolation, but the proposed strict mask forces it to zero. There is also a rounding boundary: a float64 identity probe with `depth=5e-9, eps=0` produced `x=3.0000000000000004` at a nominal right edge of 3 and rejected that column. This probe deliberately disabled the cutoff; the default cutoff rejects that tiny depth earlier. The chosen bounds rule needs numerical-boundary tests as well as a name.

My preference is to preserve the existing sampling/padding behavior for geometrically eligible, numerically safe projections and specify bounds filtering separately. If a single combined sample-valid mask is preferred, its centre/footprint rule and whether false entries are always filled must be explicit.

**3. Existing warp APIs and compatibility.** I agree with extending the existing APIs and sharing their safety machinery. I would remove the separate full public warp implementation from this proposal. The same integration milestone should cover both `warp_frame_depth` and `DepthWarper`, preserving their existing defaults and their different camera interfaces; the latter must retain its two-camera/full-intrinsics behavior.

There is one choice I would like to make explicit: should `return_mask=True` only request diagnostics, or also opt into new filling and precision behavior? I prefer separating return structure from the invalid-pixel policy, with legacy behavior as the default. If a single opt-in flag is preferred, I would document and test all of its behavioral effects. A private fixed-return helper can share computation in either design. Script/export support needs tests for the actual fixed configurations; Python typing alone would not establish it.

**4. Gradients and evidence.** The intended contract is zero contribution from each discarded output, while valid outputs may still contribute gradients to shared intrinsics, pose and source pixels. The mask is a discrete decision. Unsafe operands must be replaced before the differentiable arithmetic and sampler, consistent with [PyTorch's autograd guidance](https://docs.pytorch.org/docs/2.14/notes/autograd.html#division-by-zero-in-autograd).

Local CPU probes on torch 2.5.1 and 2.14 confirmed the ordinary missing/negative/nonfinite-depth guards, but also exposed an unfinished part: with finite `fx=1e-20`, finite rays and entirely missing depth, output is zero yet the live unprojection prefix produces a NaN intrinsics gradient. That prefix needs guarding before claiming the invalid-output gradient contract. Separately, valid half-precision cases at very small depth can overflow the pose gradient when it is cast back to half; geometric validity cannot promise representable derivatives at every scale. These bounded probes are not a complete device or export validation.

I propose first settling and testing the standalone pinhole predicate, then integrating both warps with agreed bounds/fill semantics and unchanged defaults. Sensors and LoFTR's depth-consistency policy would be separate follow-ups. Acceptance would include the missing-depth example in this issue across both warps, nontrivial cameras, boundary behavior, shared-input gradients, the supported torch floor, and export/runtime checks.

Would you prefer separate geometric and bounds information, or a combined sample mask? And should invalid filling be an independent opt-in, or explicitly part of `return_mask=True`? I would resolve those choices here before revising the implementation.

AI assistance: Codex helped inspect the sources, execute the local probes, and draft this proposal.

**Comment by ducha-aiki** [comment]
Another case for this discussion: `reproject_disparity_to_3D` (both the module-level function and `StereoCamera.reproject_disparity_to_3D`). A zero disparity (no stereo match, or a point at infinity) gives `W = 0`. The masked homogeneous divide (the #4267 rule) then returns the numerator. On a real rig that is a finite point **behind** the camera, and nothing flags it as invalid:

```python
import torch
from kornia.geometry.camera.stereo import StereoCamera

left = torch.tensor([[[100.0, 0, 4, 0], [0, 100, 3, 0], [0, 0, 1, 0]]]); right = left.clone(); right[0, 0, 3] = -50.0
print(StereoCamera(left, right).reproject_disparity_to_3D(torch.zeros(1, 1, 1, 1)).tolist())  # [[[[200.0, 150.0, -5000.0]]]]
```

The overall sign of `Q` normally cancels in the divide. At `W = 0` it does not, so `Q` and `-Q` return opposite points. For comparison, `cv2.reprojectImageTo3D` gives `inf` there, or `Z = 10000` with `handleMissingValues=True`.

#4790 documents this on `StereoCamera` and pins it as `test_wart_reproject_disparity_zero_disparity_returns_the_numerator_4555`. Whatever validity signal this issue settles on should cover this call, and that pin should then be updated.

Posted on behalf of @ducha-aiki by Claude (Opus 5.5).

**Comment by NaCr05** [comment]
Thanks for adding this case and clarifying the coordinate-policy boundary in [#4267](https://github.com/kornia/kornia/issues/4267#issuecomment-5797263613). I agree that the validity design should cover both stereo entry points while preserving the existing coordinate-return policy.

I reproduced the example locally on CPU with torch 2.5.1 and 2.14, in float32 and float64. It also shows why applying my proposed `projection_valid_mask` to the returned XYZ is insufficient: with zero disparity, `Q` returns `Z=-5000`, while `-Q` returns `Z=5000`. Both took the no-divide branch, but the latter passes a post-hoc `z > eps` check.

For stereo, I propose separating:

- **Reconstruction validity:** whether the operation produced finite, non-placeholder Euclidean coordinates. This requires the actual homogeneous `W` before division, consistency with the converter’s cutoff and working precision, and a finite result in the returned dtype.
- **Pinhole eligibility:** whether a successfully reconstructed point lies in the allowed front-of-camera region. The camera-frame `z > eps` predicate remains useful as an additional filter.

The first check should use the actual `W`, rather than `disparity != 0` or `W > 0`: a supplied `Q` can have nonzero `Q[3,3]`, and valid positive-depth points in this rig have negative `W`. The mask should agree for `Q` and `-Q`; the existing absolute cutoff’s dependence on Q scaling would be documented and tested.

I propose adding both stereo entry points to the consumer-integration stage alongside the two depth warps. Tests would cover your example, ordinary valid points, `Q`/`-Q`, offset Q matrices, cutoff boundaries, nonfinite inputs and invalid-output gradient contributions. I would also update the #4790 documentation/test, retaining legacy-output assertions where the defaults remain unchanged.

Does separating reconstruction validity from front-of-camera filtering match your intended stereo contract? I’ll keep #4701 in Draft and wait for agreement on this and the remaining mask/fill choices before revising the implementation.

AI assistance: Codex helped inspect the sources, run the local probes, and draft this reply.
